frame/mixnet/src/lib.rs

```diff
@@ -0,0 +1,675 @@
+		block_number: BlockNumberFor<T>,
+		mixnode: &Mixnode,
+	) -> bool {
+		let (phase, block_in_phase) = Self::session_phase(block_number);
+		if (phase != SessionPhase::DisconnectFromPrev) ||
+			(block_in_phase < T::NumRegisterStartSlackBlocks::get())
+		{
+			return false
+		}
+
+		let (Some(end_block), _weight) =
+			T::NextSessionRotation::estimate_next_session_rotation(block_number)
+		else {
+			// Things aren't going to work terribly well in this case as all the authorities will
+			// just pile in at the start of each register phase...
+			return true
+		};
+
+		let remaining_blocks = end_block
+			.saturating_sub(block_number)
+			.saturating_sub(T::NumRegisterEndSlackBlocks::get());
+		if remaining_blocks.is_zero() {
+			// Into the slack time at the end of the session. Not necessarily too late;
+			// registrations are accepted right up until the session ends.
+			return true
+		}
+
+		// Want uniform distribution over the remaining blocks, so pick this block with probability
+		// 1/remaining_blocks. maybe_register may be called multiple times per block; ensure the
+		// same decision gets made each time by using a hash of the block number and the mixnode's
+		// public key as the "random" source. This is slightly biased as remaining_blocks most
+		// likely won't divide into 2^64, but it doesn't really matter...
+		let random = twox(block_number, &mixnode.kx_public);
+		(random % remaining_blocks.try_into().unwrap_or(u64::MAX)) == 0
+	}
+
+	fn next_local_authority() -> Option<(AuthorityIndex, AuthorityId)> {
+		// In the case where multiple local IDs are in the next authority set, we just return the
+		// first one. There's (currently at least) no point in registering multiple times.
+		let mut local_ids = AuthorityId::all();
+		local_ids.sort();
+		NextAuthorityIds::<T>::iter().find(|(_index, id)| local_ids.binary_search(id).is_ok())
+	}
+
+	/// `session_index` must be the index of the current session. `authority_index` is the
+	/// authority index in the _next_ session. Should only be called during the register phase
+	/// (`DisconnectFromPrev`).
+	fn already_registered(session_index: SessionIndex, authority_index: AuthorityIndex) -> bool {
+		with_mixnodes!(
+			session_index,
+			false, // Registration is for the _following_ session
+			Mixnodes,
+			Mixnodes::<T>::contains_key(authority_index)
+		)
+	}
+
+	/// Try to register a mixnode for the next session. Returns `true` if a registration extrinsic
+	/// is submitted. `session_index` should match `session_status().current_index`; if it does
+	/// not, `false` is returned immediately. This should be called every block during the session,
+	/// except, after `true` is returned, this should not be called for a few blocks, to give the
+	/// submitted extrinsic a chance to get included.
+	pub fn maybe_register(session_index: SessionIndex, mixnode: Mixnode) -> bool {
+		let current_session_index = T::ValidatorSet::session_index();
+		if session_index != current_session_index {
```

Comment:
It's just to avoid needing two calls into the runtime every block (one to check the session index and one to register). I can drop it if you'd prefer the two calls.